src/battle_util.c

```diff
@@ -6209,7 +6173,7 @@ u8 AbilityBattleEffects(u8 caseID, u8 battler, u16 ability, u8 special, u16 move
         }
         break;
     case ABILITYEFFECT_ON_WEATHER: // For ability effects that activate when the battle weather changes.
-        battler = gBattlerAbility = gBattlerAttacker = gBattleScripting.battler;
+        battler = gBattlerAbility = gBattleScripting.battler;
         switch (GetBattlerAbility(battler))
```

Comment:
I think these will need a `weatherAbilityDone` check. Looking at the code, it seems like if you switch in a mon with Protosynthesis when sun is active it would constantly trigger.

Reply:
> I think these will need a weatherAbilityDone check. Looking at the code, it seems like if you switch in a mon with Protosynthesis when sun is active it would constantly trigger.

I actually tested that earlier with Castform's form change, and it didn't trigger a loop.

I tested the activation on the first turn, and also the activation after a switch out, @ghoulslash#3839.

In fact, that's exactly why I brought up earlier on the Discord server how it's weird that I need a `terrainAbilityDone` for terrain abilities to prevent them from getting stuck in a loop, while weather abilities work just fine :eyes:

I'll re-test with Protosynthesis and apply the change if it's truly needed though.

Reply:
I assume that's because `TryWeatherFormChange` would return 0 if there's no form change to do, so no script would execute. The 'weatherAbilityDone' isn't needed if every ability check has its own 'check' to make sure the associated script wouldn't run.